File: packages/mobile/lib/controllers/user-controller.dart
import 'dart:convert';

import 'package:mobile/classes/achievements.dart';
import 'package:mobile/classes/game-history.dart';
import 'package:mobile/classes/server-action.dart';
import 'package:mobile/classes/user.dart';
import 'package:mobile/environments/environment.dart';
import 'package:mobile/locator.dart';
import 'package:mobile/services/client.dart';

class UserController {
  UserController._privateConstructor();

  static final UserController _instance = UserController._privateConstructor();

  factory UserController() {
    return _instance;
  }

  final String endpoint = "${Environment().config.apiUrl}";
  final _http = getIt.get<PersonnalHttpClient>().http;

  Future<PublicUser> editUser(EditableUserFields edits) async {
    return PublicUser.fromJson(jsonDecode((await _http
            .patch(Uri.parse("$endpoint/users"), body: jsonEncode(edits)))
        .body));
  }

  Future<UserStatistics> getUserStatistics() async {
    return UserStatistics.fromJson(jsonDecode(
        (await _http.get(Uri.parse("$endpoint/users/statistics"))).body));
  }

  Future<UserSearchResult> getProfileByUsername(String username) async {
    return UserSearchResult.fromJson(jsonDecode(
        (await _http.get(Uri.parse("$endpoint/users/profile/$username")))
            .body));
  }

  Future<List<UserSearchItem>> searchUsers(String? query) async {
    return UserSearchItem.fromJsonList(jsonDecode(
        (await _http.get(Uri.parse("$endpoint/users/search?q=$query"))).body));
  }

  Future<List<RatedUser>> getRatingLeaderboard() async {
    return RatedUser.fromJsonList(jsonDecode(
        (await _http.get(Uri.parse("$endpoint/users/ratings"))).body));
  }

  Future<List<GameHistory>> getGameHistory() async {
    return GameHistory.fromJsonList(jsonDecode(
        (await _http.get(Uri.parse("$endpoint/gameHistories"))).body));
  }

  Future<List<ServerAction>> getServerActions() async {
    return ServerAction.fromJsonList(jsonDecode(
        (await _http.get(Uri.parse("$endpoint/server-actions"))).body));
  }

  Future<List<UserAchievement>> getAchievements() async {
    return UserAchievement.fromJsonList(jsonDecode(
        (await _http.get(Uri.parse("$endpoint/users/achievements"))).body));
  }
}

File: packages/client/src/app/controllers/user-controller/user.controller.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { AbstractController } from '@app/controllers/abstract-controller';
import { Observable } from 'rxjs';
import { PublicUserStatistics } from '@common/models/user-statistics';
import { GameHistoryForUser } from '@common/models/game-history';
import { PublicServerAction } from '@common/models/server-action';
import { EditableUserFields, PublicUser, RatedUser } from '@common/models/user';
import { UserSearchItem, UserSearchResult } from '@common/models/user-search';
import { UserAchievement } from '@common/models/achievement';

@Injectable({
    providedIn: 'root',
})
export class UserController extends AbstractController {
    constructor(private readonly http: HttpClient) {
        super('/');
    }

    editUser(edits: EditableUserFields): Observable<PublicUser> {
        return this.http.patch<PublicUser>(this.url('/users'), edits);
    }

    searchUsers(query: string): Observable<UserSearchItem[]> {
        return this.http.get<UserSearchItem[]>(this.url('/users/search', { q: query }));
    }

    getProfileByUsername(username: string): Observable<UserSearchResult> {
        return this.http.get<UserSearchResult>(this.url(`/users/profile/${username}`));
    }

    getUserStatistics(): Observable<PublicUserStatistics> {
        return this.http.get<PublicUserStatistics>(this.url('/users/statistics'));
    }

    getRatingLeaderboard(): Observable<RatedUser[]> {
        return this.http.get<RatedUser[]>(this.url('/users/ratings'));
    }

    getGameHistory(): Observable<GameHistoryForUser[]> {
        return this.http.get<GameHistoryForUser[]>(this.url('/gameHistories'));
    }

    getServerActions(): Observable<PublicServerAction[]> {
        return this.http.get<PublicServerAction[]>(this.url('/server-actions'));
    }

    getAchievements(): Observable<UserAchievement[]> {
        return this.http.get<UserAchievement[]>(this.url('/users/achievements'));
    }
}
